tutorials/pubsub-spring-dedup-messages/index.md

```diff
@@ -0,0 +1,244 @@
+---
+title: Deduplicate Pub/Sub Messages Using Dataflow in A Spring Boot Application
+description: Use Dataflow's PubsubIO to deduplicate Pub/Sub messages in a Spring Boot application.
+author: anguillanneuf
+tags: PubSub, Spring, Spring Cloud GCP, Dataflow, Java
+date_published: 2020-02-29
+---
+
+
+Tianzi Cai | Developer Programs Engineer | Google Cloud
+
+
+## Objectives
+- Configure a Spring Boot application to use [Cloud Pub/Sub] as a message broker.
+- Use [Cloud Dataflow] to deduplicate messages.
+
+## Before You Begin
+
+1. Install the [Cloud SDK].
+
+1. Create a new Google Cloud project via the
+   [*New Project* page],
+   or via the `gcloud` command line tool.
+
+   ```shell script
+   export PROJECT_NAME=your-google-cloud-project-id
+   gcloud projects create $PROJECT_NAME
+   ```
+
+1. [Enable billing].
+
+1. Setup the Cloud SDK to your GCP project.
+
+   ```shell script
+   gcloud init
+   ```
+
+1. [Enable the APIs](https://console.cloud.google.com/flows/enableapi?apiid=dataflow,compute_component,pubsub): Dataflow, Compute Engine, Pub/Sub.
+
+1. Create a service account JSON key via the
+   [*Create service account key* page],
+   or via the `gcloud` command line tool.
+   Here is how to do it through the *Create service account key* page.
+
+   * From the **Service account** list, select **New service account**.
+   * In the **Service account name** field, enter a name.
+   * From the **Role** list, select **Project > Owner**.
+   * Click **Create**. A JSON file that contains your key downloads to your computer.
+
+   Alternatively, you can use `gcloud` through the command line.
+
+   ```shell script
+   export PROJECT_NAME=$(gcloud config get-value project)
+   export SA_NAME=spring_app
+   export IAM_ACCOUNT=$SA_NAME@$PROJECT_NAME.iam.gserviceaccount.com
+
+   # Create the service account.
+   gcloud iam service-accounts create $SA_NAME --display-name $SA_NAME
+
+   # Set the role to Project Owner (*).
+   gcloud projects add-iam-policy-binding $PROJECT_NAME \
+     --member serviceAccount:$IAM_ACCOUNT \
+     --role roles/owner
+
+   # Create a JSON file with the service account credentials.
+   gcloud iam service-accounts keys create path/to/your/credentials.json \
+     --iam-account=$IAM_ACCOUNT
+   ```
+
+   > *Note:* The **Role** field authorizes your service account to access resources.
+   > You can view and change this field later by using the
+   > [GCP Console IAM page].
+   > If you are developing a production app, specify more granular permissions than **Project > Owner**.
```

Comment:
Would it be difficult to specify what roles this SA needs in the tutorial? We generally dissuade people from giving the Owner role.

Reply:
I'd leave it as Project Owner.

- There is a link in the notes beneath that tells users how to achieve granular control. 
- Say they want to check out Monitoring, or modify their Dataflow pipeline to write to GCS. Having Project Owner just seems easier.

Reply:
I'm still of the opinion that if we can be more specific, the better. My reasoning is not just about best practices, but also making sure we teach people about the appropriate permissions needed to accomplish what's needed in this tutorial (and eventually in production).

Reply:
I think while it's good to teach people what's needed in production, it becomes a distraction for the tutorial. 
For examples, choosing between `roles/storage.objectAdmin` and `roles/storage.admin`, or `roles/pubsub.admin` versus `roles/pubsub.publisher` could raise more questions, which are best answered by the individual service's IAM pages like [this](https://cloud.google.com/storage/docs/access-control/iam-roles) and [this](https://cloud.google.com/pubsub/docs/access-control). 
I don't see other tutorials do so either, including the ones on cloud site.

Reply:
Alex found me some examples that do grant granular roles. I updated the instructions.